### PR TITLE
[Perf] Lossless final layer prefill optimization for LLaMA via Single-Query Attention

### PR DESCRIPTION
**Pull request by hdahroug**
## Motivation

In causal decoder models, during prompt prefill (`ForwardMode.EXTEND`), only the terminal token $x_S$ produces logits to predict the next token $x_{S+1}$. The intermediate representations $h_{L-1, 1:S-1}$ in the final Transformer layer ($L-1$) have zero computational out-degree toward downstream token generation.

Currently, SGLang executes full $O(S^2)$ causal attention in FlashInfer and full-sequence SwiGLU MLP across all $S$ tokens in layer $L-1$, only pruning hidden states at `LogitsProcessor`. At 32k context, this runs $1.07$ billion redundant attention operations and 32,767 redundant MLP passes in the final layer.

This PR optimizes layer $L-1$ prefill losslessly:
1. $K$ and $V$ projections are computed for all prompt tokens and committed directly to `token_to_kv_pool`, ensuring future autoregressive decode steps ($S+1, S+2, \dots$) have full history.
2. Query attention in layer $L-1$ runs a $1 \times S$ Single-Query Attention row for the terminal token, bypassing the $S \times S$ causal attention matrix in FlashInfer.
3. $W_O$, post-attention RMSNorm, and SwiGLU MLP execute strictly on the single terminal token vector.
4. Output logits and tokens are bit-for-bit identical to baseline with zero loss in accuracy.

## Modifications

* `radix_attention.py` (+25 lines) — `save_kv_cache_only()` commits K, V to `token_to_kv_pool` directly
* `llama.py` (+148 lines) — Single-Query Attention (1 x S) + terminal-token MLP in layer L-1
* `logits_processor.py` (+14 lines) — `_get_pruned_states()` accepts pre-pruned [B, D] states
* `test/registered/models_e2e/test_final_layer_prefill_opt.py` (new) — registered CI test for deterministic token parity & logit tolerance

Key guards in `llama.py`:
- Checked via `can_optimize`: requires `forward_mode.is_extend()`, `not return_logprob`, `not capture_hidden_mode.is_full()`, and `not is_ragged_verify`.
- Any request requiring token logprobs, full hidden states, or speculative verification safely falls back to standard execution.
- Decode mode (`ForwardMode.DECODE`) is completely untouched.

## Accuracy Tests

Tested with `python3 -m sglang.benchmark.one_batch --correct` on `TinyLlama/TinyLlama-1.1B-Chat-v0.4`:
- **Token Parity**: 100% exact match against HuggingFace reference baseline across all generation steps.
- **Logit Tolerance**: $\max |L_{\text{opt}} - L_{\text{base}}| < 10^{-5}$ (within bfloat16 numerical precision).

### Reproduction Command
```bash
python3 -m sglang.bench_one_batch \
  --model-path NousResearch/Meta-Llama-3.1-8B-Instruct \
  --load-format dummy \
  --batch-size 1 \
  --input-len 1024 4096 16384 32768 \
  --output-len 1
```

Benchmarked on **NVIDIA A100-SXM4-80GB** with `NousResearch/Meta-Llama-3.1-8B-Instruct` (`--load-format dummy --batch-size 1 --output-len 1`):

Micro-benchmark of layer 31 attention kernel on A100:
- At 32k context: Full FlashInfer causal attention = **45.2 ms** vs. Single-Query Attention = **0.39 ms** (**115x kernel speedup**, saving ~41 ms end-to-end).

| Input Length (Tokens) | Baseline TTFT (ms) | Optimized TTFT (ms) | Delta (ms) | Speedup (%) |
| :--- | :--- | :--- | :--- | :--- |
| **1,024** | 86.89 | 87.30 | -0.41 ms | -0.47% (noise floor) |
| **4,096** | 271.70 | **267.58** | **+4.12 ms** | **+1.52%** |
| **16,384** | 1,279.78 | **1,265.61** | **+14.17 ms** | **+1.11%** |
| **32,768** | 3,195.18 | **3,154.25** | **+40.93 ms** | **+1.28%** |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation (N/A — internal engine optimization with no public API/arg changes).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #33986251579](https://github.com/sgl-project/sglang/actions/runs/33986251579)
Latest PR Test (Extra): :x: [Run #33986251423](https://github.com/sgl-project/sglang/actions/runs/33986251423)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33986251516](https://github.com/sgl-project/sglang/actions/runs/33986251516)

### CODE REVIEW

**Comment by hdahroug** [comment]
Hi @merrymercy @Ying1123 @Fridge003 @ispobock @HaiShaw @ch-wan @BBuf @Edwardf0t1 ,
 Rebased onto latest main and added the registered CI test under test/registered/models_e2e/test_final_layer_prefill_opt.py.
 Could you please apply the run-ci label (/tag-and-rerun-ci) so CI can run on this PR? Happy to address any feedback once it's green. Thanks!
